### PR TITLE
Source

### PR DESCRIPTION
**Pull request by maksenius**
### Description
add source connector

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/conduitio-labs/conduit-connector-sap-hana/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,4 +2,119 @@
...
+## Source
+
+The DB source connects to the database using the provided connection and starts creating records for each table row
+and each detected change.
+
+### Configuration options
+
+| Name                        | Description                                                                                                                                                                                     | Required                                  | Example                                        |
+|-----------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------|------------------------------------------------|
+| `table`                     | The name of a table in the database that the connector should  read from, by default.                                                                                                           | **true**                                  | users                                          |
```
Comment (hariso) [inline README.md:30]: >  The name of a table in the database that the connector should  read from, by default.

The sentence appears to be unfinished or?

File: README.md [hunk 2]
```diff
@@ -2,4 +2,119 @@
...
+| `snapshot`                  | Whether or not the plugin will take a snapshot of the entire table before starting cdc mode, by default true.                                                                                   | false                                     | false                                          |
+| `batchSize`                 | Size of rows batch. By default is 1000.                                                                                                                                                         | false                                     | 100                                            |
+| `auth.mechanism`            | Mechanism type of auth. Valid types: DSN, Basic, JWT, X509. By default is DSN.                                                                                                                  | false                                     | DSN                                            |
```
Comment (hariso) [inline README.md:35]: Impressive list!

File: README.md [hunk 3]
```diff
@@ -2,4 +2,119 @@
...
+| `auth.clientCertFilePath`   | Path for certification file                                                                                                                                                                     | Required for X509 type.                   | /tmp/file.cert                                 |
+| `auth.ClientKeyFilePath`    | Path for key file                                                                                                                                                                               | Required for X509 type.                   | /tmp/key.cert                                  |
+
+### Snapshot
+By default when the connector starts for the first time, snapshot mode is enabled, which means that existing data will
+be read. To skip reading existing, change config parameter `snapshot` to `false`.
+
+First time when the snapshot iterator starts work,
+it is get max value from `orderingColumn` and saves this value to position.
+The snapshot iterator reads all rows, where `orderingColumn` values less or equal maxValue, from the table in batches.
```
Comment (hariso) [inline README.md:50]: I'd rephrase this so that we don't mention the snapshot iterator (the iterator is an implementation detail and a user might become confused when reading this for the first time). Also, we should focus on the value of our documentation for users and how it might help them understand and use the connector. I believe the goal of this section is to explain what's considered a snapshot. If that is true, I'd suggest putting that all rows which exist in a table at the time the snapshot started, are considered part of snapshot.

File: README.md [hunk 4]
```diff
@@ -2,4 +2,119 @@
...
+
+
+Values in the ordering column must be unique and suitable for sorting, otherwise, the snapshot won't work correctly.
+Iterators saves last processed value from `orderingColumn` column to position to field `SnapshotLastProcessedVal`.
+If snapshot was interrupted on next start connector will parse last recorded position
+to find next snapshot rows.
```
Comment (hariso) [inline README.md:56]: This would be good as a comment in the code. For a user guide, it may be too detailed.

File: README.md [hunk 5]
```diff
@@ -2,4 +2,119 @@
...
+
+
+When all records are returned, the connector switches to the CDC iterator.
```
Comment (hariso) [inline README.md:59]: I'd be explicit and say "when all snapshot records are returned, the connector switches..."

File: README.md [hunk 6]
```diff
@@ -2,4 +2,119 @@
...
+
+Run `make build`.
+
+### Testing
+
+Run `make test` to run all the unit and integration tests.
+
+## Source
+
+The DB source connects to the database using the provided connection and starts creating records for each table row
```
Comment (hariso) [inline README.md:23]: "The SAP HANA source connector..."

File: README.md [hunk 7]
```diff
@@ -2,4 +2,119 @@
...
+to find next snapshot rows.
+
+
+When all records are returned, the connector switches to the CDC iterator.
+
+### Change Data Capture (CDC)
+
+This connector implements CDC features for DB2 by adding a tracking table and triggers to populate it. The tracking
+table has the same name as a target table with the prefix `CONDUIT_` and suffix from time when pipeline started on
+format "hhmmss". For example for table `PRODUCTS` tracking will be looks like `CONDUIT_PRODUCTS_213315`.
```
Comment (hariso) [inline README.md:65]: "tracking will be looks like" -> "the tracking table's name will be"

File: README.md [hunk 8]
```diff
@@ -2,4 +2,119 @@
...
+
+Triggers have name pattern `CD_{{TABLENAME}}_{{OPERATION_TYPE}}_{{SUFFIXNAME}}`. For example:
+`CD_PRODUCTS_INSERT_213315`
+
+
+Queries to retrieve change data from a tracking table are very similar to queries in a Snapshot iterator, but with
+`CONDUIT_TRACKING_ID` ordering column.
+
+CDC iterator periodically clears rows which were successfully applied from tracking table.
+It collects `CONDUIT_TRACKING_ID` inside the `Ack` method into a batch and clears the tracking table every 5 seconds.
```
Comment (hariso) [inline README.md:85]: IMHO we can summarize this into one sentence and say that the connector cleans up the tracking table every 5 seconds (is this configurable btw?). We don't need to go into details how that happens.

File: README.md [hunk 9]
```diff
@@ -2,4 +2,119 @@
...
+Iterator saves the last `CONDUIT_TRACKING_ID` to the position from the last successfully recorded row.
+
+If connector stops, it will parse position from the last record and will try
+to get row where `{{CONDUIT_TRACKING_ID}}` > `{{position.CDCLastID}}`.
+
+
+
+<b>Please pay attention</b>
+
+Tracking table and triggers are stayed after removing pipeline. You can clear it uses commands.
```
Comment (hariso) [inline README.md:96]: I suggest:

> The tracking table and the triggers are **not** automatically removed when a pipeline is deleted. That needs to be done manually, for example by using the following commands:

File: README.md [hunk 10]
```diff
@@ -2,4 +2,119 @@
 
 ## General
 
-The [SAP HANA](https://neo4j.com/) connector is one of Conduit plugins. It provides both, a source and a destination Neo4j connector.
+The [SAP HANA](https://www.sap.com/products/technology-platform/hana/what-is-sap-hana.html) connector is one of Conduit plugins. It provides both, a source and a
+destination Sap Hana connector.
+
+### Prerequisites
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md:8]: I have added paramgen

File: README.md [hunk 11]
```diff
@@ -2,4 +2,119 @@
...
+
+### How to build it
+
+Run `make build`.
+
+### Testing
+
+Run `make test` to run all the unit and integration tests.
+
+## Source
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md:22]: Added information

File: README.md [hunk 12]
```diff
@@ -2,4 +2,119 @@
...
+| `auth.token`                | JWT token                                                                                                                                                                                       | Required for JWT type.                    | jwt_token                                      |
+| `auth.clientCertFilePath`   | Path for certification file                                                                                                                                                                     | Required for X509 type.                   | /tmp/file.cert                                 |
+| `auth.ClientKeyFilePath`    | Path for key file                                                                                                                                                                               | Required for X509 type.                   | /tmp/key.cert                                  |
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md]: sure, fixed

File: README.md [hunk 13]
```diff
@@ -2,4 +2,119 @@
...
+Run `make test` to run all the unit and integration tests.
+
+## Source
+
+The DB source connects to the database using the provided connection and starts creating records for each table row
+and each detected change.
+
+### Configuration options
+
+| Name                        | Description                                                                                                                                                                                     | Required                                  | Example                                        |
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md]: good idea!
done

File: README.md (hunk 1 above)
Comment (maksenius) [reply README.md]: fixed

File: README.md (hunk 3 above)
Comment (maksenius) [reply README.md]: ok, I have changed it.

File: README.md (hunk 4 above)
Comment (maksenius) [reply README.md]: Removed

File: README.md (hunk 5 above)
Comment (maksenius) [reply README.md]: Added

File: README.md (hunk 6 above)
Comment (maksenius) [reply README.md]: Fixed

File: README.md (hunk 7 above)
Comment (maksenius) [reply README.md]: thanks, fixed

File: README.md (hunk 8 above)
Comment (maksenius) [reply README.md]: Changed.
This is not configurable value  like in DB2  source connector.

File: README.md (hunk 9 above)
Comment (maksenius) [reply README.md]: thanks, done

File: columntypes/columntypes.go [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+	return TableInfo{
+		ColumnTypes:   columnTypes,
+		PrimaryKeys:   primaryKeys,
+		ColumnLengths: columnLengths,
+		ColumnScales:  columnScales,
+	}, nil
+}
+
+// ConvertStructureData converts a sdk.StructureData values to a proper database types.
+func ConvertStructureData(
```
Comment (hariso) [inline columntypes/columntypes.go:210]: Tinies nitpick possible: `ConvertStructuredData` (missing `d` in `structured`)

File: columntypes/columntypes.go [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+				return nil, fmt.Errorf("marshal: %w", err)
+			}
+
+			result[key] = string(bs)
+
+			continue
+		}
+
+		// Converting value to time if it is string.
+		switch columnTypes[strings.ToLower(key)] {
```
Comment (hariso) [inline columntypes/columntypes.go:239]: Why do we do `strings.ToLower(key)`?

File: config/errors.go [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+	errRequiredUsernameParameter = errors.New("username is required parameter for basic auth")
+	// errRequiredPasswordParameter occurs when there's empty password parameter for basic auth.
+	errRequiredPasswordParameter = errors.New("password is required parameter for basic auth")
+	// errRequiredTokenParameter occurs when there's empty token parameter for jwt auth.
+	errRequiredTokenParameter = errors.New("token is required for jwt auth")
+	// errRequiredClientCertFileParameter occurs when there's empty client cert file parameter for x509 auth.
+	errRequiredClientCertFileParameter = errors.New("client cert file path is required for x509 auth")
+	// errRequiredClientCertFileParameter occurs when there's empty client key file path parameter for x509 auth.
+	errRequiredClientKeyFileParameter = errors.New("client key file path is required for x509 auth")
+)
```
Comment (hariso) [inline config/errors.go:38]: WDYT about  replacing this with a function:
```go
func requiredAuthParam(auth, param) {
    return fmt.Errorf("%s is required for %s auth", param, auth)
}
```

File: source/position/position.go [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+	// SuffixName special suffix that connector uses for identify tracking table and triggers.
+	SuffixName string
+}
+
+// ParseSDKPosition parses SDK position and returns Position.
+func ParseSDKPosition(p sdk.Position) (*Position, error) {
+	var pos Position
+
+	if p == nil {
+		return nil, nil //nolint:nilnil // it is ok to return nils on this case
```
[the remark this answers is not in the document]
Comment (maksenius) [reply source/position/position.go]: fixed, my fault

File: source/source_test.go [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+
+	t.Run("failed_next", func(t *testing.T) {
+		t.Parallel()
+
+		ctrl := gomock.NewController(t)
+		ctx := context.Background()
+
+		it := mock.NewMockIterator(ctrl)
+		it.EXPECT().HasNext(ctx).Return(true, nil)
+		it.EXPECT().Next(ctx).Return(sdk.Record{}, errors.New("key is not exist"))
```
[the remark this answers is not in the document]
Comment (maksenius) [reply source/source_test.go]: changed

File: source/iterator/iterator.go [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+	// tableInfo - general information about column types, primary keys.
+	tableInfo columntypes.TableInfo
+}
+
+// CombinedParams is an incoming params for the [NewCombinedIterator] function.
+type CombinedParams struct {
+	DB             *sqlx.DB
+	Table          string
+	OrderingColumn string
+	CfgKeys        []string
```
[the remark this answers is not in the document]
Comment (maksenius) [reply source/iterator/iterator.go:60]: it has special reason. We can get primary keys from table schema and we can set keys in config,  I decided to use this name to better distinguish

File: helper/db.go [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+		return sqlx.NewDb(sql.OpenDB(con), driverName), nil
+	case config.X509AuthType:
+		con, err := driver.NewX509AuthConnectorByFiles(c.Host, c.ClientCertFilePath, c.ClientKeyFilePath)
+		if err != nil {
+			return nil, fmt.Errorf("new X509 auth: %w", err)
+		}
+
+		return sqlx.NewDb(sql.OpenDB(con), driverName), nil
+	default:
+		return nil, config.ErrInvalidAuthMechanism
```
[the remark this answers is not in the document]
Comment (maksenius) [reply helper/db.go]: good idea, done

File: source/iterator/iterator.go [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+		}
+
+		c.keys = cfgKeys
+
+		return
+	}
+
+	// second priority primary keys from table.
+	if len(c.keys) > 0 {
+		return
```
Comment (hariso) [inline source/iterator/iterator.go:240]: Where exactly do we set the keys to be primary keys from the table?

File: columntypes/columntypes.go (hunk 14 above)
Comment (maksenius) [reply columntypes/columntypes.go]: fixed

File: columntypes/columntypes.go [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+	rows, err := querier.QueryContext(ctx, queryIfTableExist, tableName)
+	if err != nil {
+		return TableInfo{}, fmt.Errorf("execute query exist table: %w", err)
+	}
+
+	defer rows.Close() //nolint:staticcheck,nolintlint
+
+	for rows.Next() {
+		var count int
+		er := rows.Scan(&count)
```
[the remark this answers is not in the document]
Comment (maksenius) [reply columntypes/columntypes.go]: yes it is looks better , done

File: columntypes/columntypes.go [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+	ColumnTypes map[string]string
+	// PrimaryKeys - primary keys column names.
+	PrimaryKeys []string
+	// ColumnLengths - column name with length.
+	ColumnLengths map[string]int
+	// ColumnScales - column name with scale.
+	ColumnScales map[string]*int
+}
+
+// GetColumnQueryPart prepare query part about creation column for tracking table.
```
[the remark this answers is not in the document]
Comment (maksenius) [reply columntypes/columntypes.go:92]: added example

File: source/position/position.go [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+	// SnapshotLastProcessedVal - last processed value from ordering column.
+	SnapshotLastProcessedVal any
+	// SnapshotMaxValue - max value from ordering column.
+	SnapshotMaxValue any
+
+	// CDC information.
+	// CDCLastID - last processed id from tracking table.
+	CDCLastID int
+	// SuffixName special suffix that connector uses for identify tracking table and triggers.
+	SuffixName string
```
Comment (hariso) [inline source/position/position.go:47]: Maybe we can use the full tracking table name?

File: config/errors.go (hunk 16 above)
Comment (maksenius) [reply config/errors.go]: done

File: source/iterator/cdc.go [hunk 25]
```diff
@@ -0,0 +1,487 @@
...
+			sdk.StructuredData(keysMap), sdk.RawData(transformedRowBytes)), nil
+	case updateOperation:
+		return sdk.Util.Source.NewRecordUpdate(convertedPosition, metadata,
+			sdk.StructuredData(keysMap), nil, sdk.RawData(transformedRowBytes)), nil
+	case deleteOperation:
+		return sdk.Util.Source.NewRecordDelete(convertedPosition, metadata,
+			sdk.StructuredData(keysMap)), nil
+	default:
+		return sdk.Record{}, ErrUnknownOperatorType
+	}
```
Comment (hariso) [inline source/iterator/cdc.go:223]: Same as in other places: IMHO it's more natural that the payload for DB sources comes in the form of `sdk.StructuredData`.

File: source/iterator/cdc.go [hunk 26]
```diff
@@ -0,0 +1,487 @@
...
+			i.tableSrv.close()
+
+			err := i.db.Close()
+			if err != nil {
+				return fmt.Errorf("close db:%w", err)
+			}
+
+			return nil
+		}
+	}
```
Comment (hariso) [inline source/iterator/cdc.go:261]: The only difference I see between these two cases is `return nil` which we have when the timeout is reached.

Maybe we can simplify the switch as:
```go
select {
    	case <-i.tableSrv.canCloseCh:
    	case <-time.After(waitingTimeoutSec * time.Second):
            // close tableSrv
}
```

File: source/iterator/snapshot.go [hunk 27]
```diff
@@ -0,0 +1,261 @@
...
+	// position last recorded position.
+	position *position.Position
+	// columnTypes column types from table.
+	columnTypes map[string]string
+	// suffixName special suffix that connector uses for identify tracking table and triggers.
+	suffixName string
+}
+
+type snapshotParams struct {
+	db             *sqlx.DB
```
[the remark this answers is not in the document]
Comment (maksenius) [reply source/iterator/snapshot.go:64]: issue was created
https://github.com/conduitio-labs/conduit-connector-sap-hana/issues/15

File: source/iterator/snapshot.go [hunk 28]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+// HasNext check ability to get next record.
+func (i *snapshotIterator) HasNext(ctx context.Context) (bool, error) {
+	if i.rows != nil && i.rows.Next() {
+		return true, nil
+	}
+
+	if err := i.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
```
[the remark this answers is not in the document]
Comment (maksenius) [reply source/iterator/snapshot.go:115]: if we get error inside this function we get  "has next" in error message. Because it was wrapped
```
	switch {
	case c.snapshot != nil:
		hasNext, err := c.snapshot.HasNext(ctx)
		if err != nil {
			return false, fmt.Errorf("snapshot has next: %w", err)
		}
```
so we can something like it:
```
source has next: snapshot has next: load rows: some error
```

File: source/iterator/snapshot.go [hunk 29]
```diff
@@ -0,0 +1,261 @@
...
+	i.position = &pos
+
+	metadata := sdk.Metadata(map[string]string{metadataTable: i.table})
+	metadata.SetCreatedAt(time.Now())
+
+	return sdk.Util.Source.NewRecordSnapshot(
+			sdkPos,
+			metadata,
+			sdk.StructuredData(keysMap),
+			sdk.RawData(transformedRowBytes)),
```
[the remark this answers is not in the document]
Comment (maksenius) [reply source/iterator/snapshot.go:177]: because conduit can return error if we structured data and  value with time.Time type in payload
Something like it
```
source stream was stopped unexpectedly: error reading from source: read stream error: error converting key: proto: invalid type: time.Time
```

File: README.md [hunk 30]
```diff
@@ -2,4 +2,109 @@
 
 ## General
 
-The [SAP HANA](https://neo4j.com/) connector is one of Conduit plugins. It provides both, a source and a destination Sap Hana connector.
+The [SAP HANA](https://www.sap.com/products/technology-platform/hana/what-is-sap-hana.html) connector is one of Conduit plugins. 
+It provides the source Sap Hana connector.
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md:6]: fixed

File: helper/db.go [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+		return sqlx.NewDb(sql.OpenDB(con), driverName), nil
+	case config.X509AuthType:
+		con, err := driver.NewX509AuthConnectorByFiles(c.Host, c.ClientCertFilePath, c.ClientKeyFilePath)
+		if err != nil {
+			return nil, fmt.Errorf("new X509 auth: %w", err)
+		}
+
+		return sqlx.NewDb(sql.OpenDB(con), driverName), nil
+	default:
+		return nil, fmt.Errorf("%w:%s", config.ErrInvalidAuthMechanism, c.Mechanism)
```
[the remark this answers is not in the document]
Comment (maksenius) [reply helper/db.go:57]: fixed

File: source/iterator/snapshot.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/snapshot.go:177]: There it says "error converting key", and not payload though? We use structured data in Postgres as well (and that _should_ work).

File: source/iterator/snapshot.go [hunk 32]
```diff
@@ -0,0 +1,261 @@
...
+
+	return nil
+}
+
+// LoadRows selects a batch of rows from a database, based on the CombinedIterator's
+// table, columns, orderingColumn, batchSize and the current position.
+func (i *snapshotIterator) loadRows(ctx context.Context) error {
+	builder := sqlbuilder.NewSelectBuilder()
+
+	builder.Select("*")
```
[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/snapshot.go:217]: To be honest, I can't quite remember that, that's very possible. I would say that transforms provide the possibility, but they are not optimal.

[the remark this answers is not in the document]
Comment (maksenius) [reply source/iterator/snapshot.go:217]: ok I will create feature for it

File: source/iterator/snapshot.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (maksenius) [reply source/iterator/snapshot.go:177]: ok, I will retest it

File: source/position/position.go (hunk 24 above)
Comment (maksenius) [reply source/position/position.go:47]: I was thinking about , but SuffixName is shorter  than tracking table name. For example for table name `SOME_REALLY _LONG_TABLE_NAME` tracking table name will be `COUNDUIT_SOME_REALLY _LONG_TABLE_NAME_203907` and suffix will be `203907`

File: source/iterator/cdc.go (hunk 25 above)
Comment (maksenius) [reply source/iterator/cdc.go:223]: please check it
https://github.com/conduitio-labs/conduit-connector-sap-hana/pull/12#discussion_r1097502451

File: source/iterator/iterator.go (hunk 21 above)
Comment (maksenius) [reply source/iterator/iterator.go:242]: I decided to move this logic to here, it will be more clear

File: source/iterator/cdc.go (hunk 26 above)
Comment (maksenius) [reply source/iterator/cdc.go:261]: For us is valid case when we get signal from this channel tableSrv.canCloseCh, because it is means that connector clears tracking table and we can safely close it. We use timer for purpose if clearing tracking process stuck

File: README.md [hunk 33]
```diff
@@ -2,4 +2,109 @@
...
+and each detected change. It supports gets snapshot from table and catches CDC (Change Data Captured) changes.
+
+### Configuration options
+
+| Name                      | Description                                                                                                                                                                                     | Required                                   | Example                                           | By default |
+|---------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------|---------------------------------------------------|------------|
+| `table`                   | The name of a table in the database that the connector should read from.                                                                                                                        | **true**                                   | users                                             |            |
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md:31]: Not for now
It has issues with get table info functionality, and it is cause problem with transform logic. Also it has problems with CDC

File: source/iterator/snapshot.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (maksenius) [reply source/iterator/snapshot.go:217]: https://github.com/conduitio-labs/conduit-connector-sap-hana/issues/16

File: source/iterator/snapshot.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/snapshot.go:177]: This is quite unexpected...

File: columntypes/columntypes.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (hariso) [reply columntypes/columntypes.go:92]: Awesome!

File: README.md (hunk 33 above)
[the remark this answers is not in the document]
Comment (hariso) [reply README.md:31]: Great, thanks for the clarification! If you could mention that in the readme (that the connector supports tables) that would be great!

File: source/position/position.go (hunk 24 above)
Comment (maksenius) [reply source/position/position.go:47]: it is not big deal, I am going to change it

File: README.md (hunk 33 above)
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md:31]: We have sentence about it
```
The SAP HANA source connects to the database using the provided connection and starts creating records for each table row
and each detected change.
```

File: source/iterator/cdc.go (hunk 26 above)
Comment (maksenius) [reply source/iterator/cdc.go]: Added comments and logs

File: README.md (hunk 33 above)
[the remark this answers is not in the document]
Comment (hariso) [reply README.md:31]: My bad, didn't read it thoroughly!

File: source/iterator/snapshot.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/snapshot.go:177]: Confirmed it's an issue in Conduit: https://github.com/ConduitIO/conduit/issues/866

**Review by hariso — APPROVED** [review]
Great work! :1st_place_medal:
